lib/linked_list.rb

```diff
@@ -12,152 +11,316 @@ def initialize(value, next_node = nil)
 
 # Defines the singly linked list
 class LinkedList
-    def initialize
-      @head = nil # keep the head private. Not accessible outside this class
-    end
-
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def add_first(value)
-      raise NotImplementedError
-    end
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
 
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    # Time Complexity:  
-    # Space Complexity
-    def search(value)
-      raise NotImplementedError
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
+  def add_first(value)
+    if self.length != 0

```

Comment:
You can do this without calling the `length` method since it doesn't matter how long the list is to add an element to the front.

Also since you're doing `.length`, this causes your algorithm to be O(n) time complexity.